### DISCUSSION TITLE
Supporting Vendor-Specific Display Controllers

### DISCUSSION THREAD

**Discussion by ioannis-karachalios** [zephyrproject-rtos/zephyr#63051] [General] [answered]
Hi community,

I am a SW engineer from Renesas Electronics, and we are currently evaluating the possibility to support the display controller integrated in our Bluetooth Low Energy, high-end devices. For starters, we would be interested in demonstrating the display controller of the DA1469x SoC which is currently supported by Zephyr. We have done some research on this topic and our understanding is that there is no dedicated driver framework when it comes to interfacing with display drivers. Instead, a more generic approach is adopted using a 4-wire or 3-wire SPI interface. Thing is that our display controller is what differentiates us from other competitors in the market when it comes to developing low-power, wireless, MCU-based wearable applications. Following is some of the advantages for one to utilize our display controller instead of adopting the mentioned generic approach.

- Timing generator capable of supporting several different serial and parallel display interfaces for instance, MIPI DPI (VSYNC/HSYNC), MIPI DBI-C (SPI3/4), serial Sharp and serial/parallel JDI, to name a few.

- DMA accelerator with geometry awareness. One would argue that this accelerator could be replaced by a generic DMA channel, but this should not work for all the cases. For example, when the frame buffer is stored in an external medium, for instance PSRAM/FLASH, and the output pixel rate is higher than the input pixel rate then a mechanism, namely pixels pre-fetching, must be employed. Otherwise, the display controller will fail to execute a frame update.

- Pixel formatter used to perform on-the-fly and zero-latency color formats conversions while data are being fetched from the frame buffer. For instance, frame buffer's pixels data with RGB565 color format could be converted to RGB888 at the output.

- Layer overlay circuitry capable of performing geometry tasks for instance moving the frame buffers onto the X/Y Cartesian coordinate system, performing partial updates, color blending etc. 

Having mentioned all the above, one can conclude that our display controller should offload the main processor (ARM Cortex-M33) and contribute to lower power consumption required for battery operated wearable products.

Our main purpose of this first communication is to kick off discussions on the possibilities to port our display controller to the Zephyr project. For instance, could we create an intermediate interface between Zephyr's generic display framework and our device-specific display accelerator, or could we initiate discussions on exploring a new accelerator-oriented display driver framework? 

Thanks in advance.

Best regards,
Ioannis Karachalios

**Comment by DerekSnell** [comment]
Hi @ioannis-karachalios ,
NXP is working on a similar framework for Zephyr.  Please see [[RFC] 2D Accelerator API Proposal](https://github.com/zephyrproject-rtos/zephyr/pull/60840), and we look forward to your feedback there.

Best regards

**Reply by ioannis-karachalios** [reply]
Hi @DerekSnell,

We thoroughly studied the mentioned framework and we concluded that the proposed solution should not meet our controller's peculiarities simply because our display accelerator does not exhibit any rendering capabilities as opposed to vector accelerators. 

Best regards

**Reply by danieldegrasse** [reply, marked as answer]
Hi @ioannis-karachalios- thank you for the initial feedback. I would appreciate you engaging on the RFC, if you are willing. NXP has similar 2D accelerators which are only capable of operations such as pixel format conversion or display rotation. The current state of that RFC is that several community members have expressed concern that a generic API will limit performance of some accelerators, so we may explore device-specific interfaces for each accelerator

**Reply by ioannis-karachalios** [reply]
Hi @danieldegrasse, 
Our concern is that these common functionalities, e.g. pixel format conversion, can only be performed while data pixels are being fetched from a frame buffer and outputted through the selected display interface e.g. 4-wire SPI. It's not that our accelerator supports read-write-modify or memory-to-memory operations. Nevertheless, I will also engage on the RFC and provide a first feedback. 
 
Best regards

**Reply by danieldegrasse** [reply]
Understood- we do have accelerators like this as well, that can only convert data while it in flight to the display. This is what the `accel2d_write_display` API is intended to cover- if the source buffer format did not match the display format, the color format conversion would be performed by the accelerator on the fly.

**Reply by ioannis-karachalios** [reply]
Indeed, this should accommodate our accelerator's capabilities. I was wandering whether the proposed RFC, and apart from dealing with rendering operations (on-the-fly or in-memory) will also take care of display interfaces. Will there be any interface layer where one can employ standardized or even proprietary display interface protocols? Typically this should include I/O definitions, bus definitions and driver settings definitions.

**Reply by ioannis-karachalios** [reply]
Hi @danieldegrasse could you please answer to my last question on the display support? It's not clear to me how `display_write` will actually work. Thanks in advance.

**Reply by danieldegrasse** [reply]
Hi @ioannis-karachalios, apologies- the proposed operation would indeed be expected to replace `display_write`, so it would be able to handle standard/propriety interface protocols in an SOC specific way. You may also want to look at #64587. This should be able to support SPI 3/4 wire display controllers

**Reply by ioannis-karachalios** [reply]
Many thanks for bringing this driver class to life. I will follow up in the corresponding PR.

**Reply by ioannis-karachalios** [reply]
@danieldegrasse I have already provided some feedback in the corresponding PR but I have another question which I'd prefer to discuss it here.
Is there is any provision or previous discussions for the layer concept as some LCD controllers natively support a single or multiple layers. For instance our DA1469x SoC supports a single layer which will not really be used here but we are also evaluating other devices that support multiple layers. I guess this will require changes in the display driver API so one can select the target layer along with some layer properties e.g. position, transparency etc.

**Reply by danieldegrasse** [reply]
Hi @ioannis-karachalios, there is no provision currently, no. I am aware of LCD controllers like this, we also have some at NXP. It would probably require a change to the display driver API, but I'm not sure LVGL can currently leverage layers. Without a way to use layers with a graphics framework, I'm not sure how much performance improvement we will see. I *believe* LVGLv9 may have a layer concept, but that rework is still in progress.

**Reply by ioannis-karachalios** [reply]
Hi @danieldegrasse, understood. Thanks for the feedback.
